CEP-28: Implement Bulk API endpoints and introduce the Sidecar Client to Support Bulk Analytics

frankgh opened the pull request:
This PR implements the remaining endpoints needed to perform Bulk Analytics operations that allow
reading and writing data from Cassandra in Bulk. The new endpoints include:

- Endpoint to create snapshots
- Endpoint to clear a snapshot
- Endpoint to upload SSTable components
- Endpoint to clean up uploads for SSTable components
- Endpoint to import SSTable components
- Endpoint to retrieve gossip info
- Endpoint to retrieve the time skew for the server
- Endpoint to retrieve the ring information

Sidecar Client

Introduces the fully featured sidecar client to access Cassandra Sidecar endpoints.
It offers support for retries and Sidecar instance selection policies. The client
project itself is technology-agnostic, but we provide a vertx implementation for
the `HttpClient`. The Sidecar vertx-client can be published as a shaded-jar to be
consumed by clients where the dependencies can cause issues, especially in environments
where the dependencies are not always controlled by the consumers (for example Spark).

Co-authored-by: Saranya Krishnakumar <saranya_k@apple.com>
Co-authored-by: Yifan Cai <ycai@apache.org>
Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>
Co-authored-by: Doug Rohrer <drohrer@apple.com>

frankgh committed:
Implement Bulk API endpoints and introduce the Sidecar Client to Support Bulk Analytics

This commit implements the remaining endpoints needed to perform Bulk Analytics operations that allow
reading and writing data from Cassandra in Bulk. The new endpoints include:

- Endpoint to create snapshots
- Endpoint to clear a snapshot
- Endpoint to upload SSTable components
- Endpoint to clean up uploads for SSTable components
- Endpoint to import SSTable components
- Endpoint to retrieve gossip info
- Endpoint to retrieve the time skew for the server
- Endpoint to retrieve the ring information

Sidecar Client

Introduces the fully featured sidecar client to access Cassandra Sidecar endpoints.
It offers support for retries and Sidecar instance selection policies. The client
project itself is technology-agnostic, but we provide a vertx implementation for
the `HttpClient`. The Sidecar vertx-client can be published as a shaded-jar to be
consumed by clients where the dependencies can cause issues, especially in environments
where the dependencies are not always controlled by the consumers (for example Spark).

Co-authored-by: Saranya Krishnakumar <saranya_k@apple.com>
Co-authored-by: Yifan Cai <ycai@apache.org>
Co-authored-by: Francisco Guerrero <frank.guerrero@gmail.com>
Co-authored-by: Doug Rohrer <drohrer@apple.com>

frankgh committed:
Update circle CI for skipping the integration tests for the docker tests

frankgh committed:
Expose methods from private to protected in the SidecarClient

This allows clients that extend functionality from the client to use methods from
the base classes

frankgh committed:
Ninja fix: Use shadow publishing for vertx-client-all

This prevents the pom file to add the vertx-client as a dependency while
the classes are correctly added to the fat jar

frankgh committed:
Do not shade jackson-annotations

When shading jackson-annotations, extension libraries won't be able to annotate
POJOs with the unshaded jackson annotations (for example @JsonProperty), and
they would be forced to use the shaded annotation. In this commit, we exclude
jackson-annotations so that extension clients can keep annotating POJOs without
the need to use the shaded annotations.

dineshjoshi committed:
Add more logging

dineshjoshi committed:
Update naming

dineshjoshi commented on cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicas.java:
```diff
@@ -0,0 +1,417 @@
+        BigInteger maxValue = this.partitioner.maxToken;
+        if (this.start.compareTo(other.start) >= 0)
+        {
+            // TODO:
```
TODO?

dineshjoshi commented on vertx-client/src/test/resources/log4j2.xml:
```diff
@@ -0,0 +1,24 @@
+<Configuration status="ERROR">
+    <Properties>
+        <Property name="LOG_PATTERN">%d{yyyy-MM-dd'T'HH:mm:ss.SSSZ} %p %m%n</Property>
+        <Property name="APP_LOG_ROOT">c:/temp</Property>
```
c:/temp?

dineshjoshi commented on src/test/resources/sidecar_with_single_multiple_instances.yaml:
```diff
@@ -38,6 +38,8 @@ sidecar:
       - stream_requests_per_sec: 5000
       - delay_sec: 5
       - timeout_sec: 10
+  - allowable_time_skew_in_minutes: 60
```
Is 60 minutes a reasonable default?

frankgh replied:
this is configurable. Do you have a different number in mind to suggest?

frankgh replied:
discussed offline. we decided to leave the value as is for now.

dineshjoshi commented on src/test/java/org/apache/cassandra/sidecar/routes/sstableuploads/BaseUploadsHandlerTest.java:
```diff
@@ -0,0 +1,204 @@
+ */
+class BaseUploadsHandlerTest
+{
+    protected final Logger logger = LoggerFactory.getLogger(this.getClass());
```
logger -> LOGGER

yifan-c replied:
The `logger` is not `static`.

frankgh replied:
yeah, classes that extend this base class can use the logger.

dineshjoshi reviewed (changes requested):
Thanks @frankgh. Minor nits only in comments.

I have also taken the time to put my feedback in [this](https://github.com/dineshjoshi/cassandra-sidecar/commits/CEP-28-bulk-apis) branch. Please cherry pick the commits. Thanks.

yifan-c commented on cassandra40/build.gradle:
```diff
@@ -12,7 +13,34 @@ repositories {
+
+}
+
+publishing {
```
nit: maybe declare it in the `allprojects` block of the root project.

frankgh replied:
Will skip for now. I think we should create a JIRA to upgrade gradle and do this as part of that effort. What do you think?

yifan-c replied:
As said, feel free to ignore the nits.

frankgh committed:
Revert renaming for JMX mbean methods

frankgh committed:
Address PR feedback from Dinesh

frankgh commented:
@dineshjoshi thanks for the review. I've applied your patch and addressed your comments. Let me know if there is any additional feedback

yifan-c commented on cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/RingProvider.java:
```diff
@@ -0,0 +1,211 @@
+    }
+
+    /**
+     * Data class to get status of endppints
```
nit
```suggestion
     * Data class to get status of endpoints
```

yifan-c commented on cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/StorageJmxOperations.java:
```diff
@@ -1,12 +1,103 @@
+     * we calculate the percentage using replication factor.
+     * If Keyspace == null, this method will try to verify if all the keyspaces
+     * in the cluster have the same replication strategies and if yes then we will
+     * use the first else a empty Map is returned.
```
nit
```suggestion
     * use the first else an empty Map is returned.
```

yifan-c commented on cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicas.java:
```diff
@@ -0,0 +1,417 @@
+     * @param other the range we are currently processing to check if "this" intersects it
+     * @return true if "this" range intersects the other
+     */
+    private boolean intersects(TokenRangeReplicas other)
+    {
+        return (other.end.compareTo(partitioner.maxToken) == 0 && this.start.compareTo(other.start) > 0) ||
+               this.start.compareTo(other.end) < 0;
+    }
```
I think this implementation is wrong. You want to compare `this.end.compareTo(other.start) < 0;`. Before the comparison, it needs to sort the 2 ranges by `start`, and make sure to compare left range with right range.
~Beside that, the first condition that checks with the maxToken of the partitioner seems not necessary. It only wants to compare 2 ranges.~
(Edit: looks like wrap around is considered. In that case, either this or that range needs to unwrap to check intersection.)

The below test fails 

```java
    @Test
    void testInterests()
    {
        TokenRangeReplicas range1 = new TokenRangeReplicas(BigInteger.valueOf(1), BigInteger.valueOf(10), Partitioner.Murmur3, new HashSet<>());
        TokenRangeReplicas range2 = new TokenRangeReplicas(BigInteger.valueOf(9), BigInteger.valueOf(12), Partitioner.Murmur3, new HashSet<>());
        assertThat(range1.intersects(range2)).isTrue();
        assertThat(range2.intersects(range1)).isTrue();

        TokenRangeReplicas range3 = new TokenRangeReplicas(BigInteger.valueOf(1), BigInteger.valueOf(10), Partitioner.Murmur3, new HashSet<>());
        TokenRangeReplicas range4 = new TokenRangeReplicas(BigInteger.valueOf(11), BigInteger.valueOf(20), Partitioner.Murmur3, new HashSet<>());
        assertThat(range3.intersects(range4)).isFalse();
        assertThat(range4.intersects(range3)).isFalse();
    }
```

frankgh replied:
Adding this test and providing a fixed implementation

yifan-c reviewed (changes requested):
Feel free to ignore the nits.

yifan-c commented on cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicas.java:
```diff
@@ -0,0 +1,415 @@
+
+
+/**
+ * Representation of a token range and the corresponding mapping to replica-set hosts
```
add a sentence to explain which end of the range is exclusive. I would assume `(start, end]`

yifan-c commented on cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicas.java:
```diff
@@ -0,0 +1,415 @@
+        return start.compareTo(end) >= 0;
+    }
+
+    private boolean isSubsetOf(TokenRangeReplicas other)
```
Does it check if this range is entirely covered by the other range? In other word, the `other` range contains `this` range. 

If so, I think the implementation is wrong. See test below
```java
    @Test
    void testSubsetRelationship()
    {
        // wraps around
        TokenRangeReplicas range1 = new TokenRangeReplicas(BigInteger.valueOf(100), BigInteger.valueOf(-100), Partitioner.Murmur3, new HashSet<>());
        TokenRangeReplicas range2 = new TokenRangeReplicas(BigInteger.valueOf(100), Partitioner.Murmur3.maxToken, Partitioner.Murmur3, new HashSet<>());
        assertThat(range1.isSubsetOf(range2)).isFalse();
        assertThat(range2.isSubsetOf(range1)).isTrue();
    }
```

nit: the name could be more succinct if it is called "contains". The implementation needs to be adjusted if being renamed.

frankgh replied:
adding this test and providing a fixed implementation

yifan-c commented on cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicas.java:
```diff
@@ -0,0 +1,415 @@
+     * @param other the next range in the range list to compare
+     * @return true if "this" range is larger than the other
+     */
+    private boolean isLarger(TokenRangeReplicas other)
```
1. Please write unit tests for this method. 
2. What qualifies a range to be larger? I am still confused after reading the java doc. The first sentence says it compares the lengths, which makes sense. But the later sentences seem to indicate that it also depends on the position of the ranges.

frankgh replied:
Added some unit tests. I also added javadocs clarifying that this method assumes that the ranges are normalized and unwrapped, i.e. `this` comes before `other` AND there's no wrapping around the min token

yifan-c commented on checkstyle.xml:
```diff
@@ -139,6 +139,7 @@ page at http://checkstyle.sourceforge.net/config.html -->
 
         <module name="JavadocType">
             <property name="scope" value="protected" />
+            <property name="allowUnknownTags" value="true" />
```
👍

yifan-c commented on client/src/main/java/org/apache/cassandra/sidecar/client/RequestContext.java:
```diff
@@ -0,0 +1,455 @@
+ */
+public class RequestContext
+{
+    protected static final SchemaRequest FULL_SCHEMA_REQUEST = new SchemaRequest();
+    protected static final TimeSkewRequest TIME_SKEW_REQUEST = new TimeSkewRequest();
+    protected static final NodeSettingsRequest NODE_SETTINGS_REQUEST = new NodeSettingsRequest();
+    protected static final RingRequest RING_REQUEST = new RingRequest();
```
nit: the request objects, as well as the request builder methods below, do not feel should part of `RequestContext`, which supplies the context for a request. 
But I think it is fine to leave them as is for now.

frankgh replied:
maybe we can have it as a follow up

yifan-c commented on client/src/main/java/org/apache/cassandra/sidecar/client/SidecarConfig.java:
```diff
@@ -0,0 +1,125 @@
+        private long retryDelayMillis = DEFAULT_RETRY_DELAY_MILLIS;
+        private long maxRetryDelayMillis = DEFAULT_MAX_RETRY_DELAY_MILLIS;
+
+        public Builder()
+        {
+        }
```
It can be removed.

yifan-c commented on client/src/main/java/org/apache/cassandra/sidecar/client/SidecarInstanceImpl.java:
```diff
@@ -0,0 +1,105 @@
+     * {@inheritDoc}
+     */
+    @Override
+
```
remove the empty line.

yifan-c commented on client/src/main/java/org/apache/cassandra/sidecar/client/exception/UnexpectedStatusCodeException.java:
```diff
@@ -0,0 +1,53 @@
+     * @param statusCode the response status code
+     * @param body       the body of the response
+     */
+    public UnexpectedStatusCodeException(int statusCode, String body)
```
Is it safe to assume the body is always `String`?

frankgh replied:
good point. I think this can be a binary file that we stream

frankgh replied:
this constructor is called from 
```
    UnexpectedStatusCodeException unexpectedStatusCode(HttpResponse response)
    {
        return new UnexpectedStatusCodeException(response.statusCode(), response.contentAsString());
    }
```
and `response.contentAsString()` is a best effort attempt to convert the raw response to a UTF-8 String,

frankgh replied:
will leave for now as it's an unexpected case, what do you think?

yifan-c commented on client/src/main/java/org/apache/cassandra/sidecar/client/request/Request.java:
```diff
@@ -0,0 +1,95 @@
+     */
+    public HttpMethod method()
+    {
+        return HttpMethod.GET;
```
nit: it would read more clear to not supply a default HttpMethod, and enforce the subclasses to define it.

yifan-c commented on client/src/main/java/org/apache/cassandra/sidecar/client/retry/BasicRetryPolicy.java:
```diff
@@ -0,0 +1,297 @@
+public class BasicRetryPolicy extends RetryPolicy
+{
+    protected static final String RETRY_AFTER = "Retry-After";
+    // Use -1 to retry indefinitely
```
nit: define a constants

yifan-c commented on client/src/main/java/org/apache/cassandra/sidecar/client/retry/ExponentialBackoffRetryPolicy.java:
```diff
@@ -0,0 +1,90 @@
+    {
+        long exponentialValue = (long) Math.pow(2, attempts - 1);
+        // do not multiply times retryDelayMillis, if we've reached the Long.MAX_VALUE already to avoid overflows
+        long retryDelay = exponentialValue == Long.MAX_VALUE ? exponentialValue : exponentialValue * retryDelayMillis;
```
`exponentialValue * retryDelayMillis` could overflow and result in a negative value. 
But it will never happen in practice, no human could possibly wait for this long time.

frankgh replied:
this check here is so it doesn't overflow. Math.pow will return Long.MAX_VALUE if it will overflow.

yifan-c replied:
It will. Say `exponentialValue == Long.MAX_VALUE / 2` and `retryDelayMillis == 3`, and boom!
```shell
jshell> exponentialValue * retryDelayMillis
$10 ==> -4611686018427387907
```

frankgh replied:
ah yeah, that will overflow for sure. I will make it more robust

yifan-c commented on client/src/main/java/org/apache/cassandra/sidecar/client/selection/OrderedInstanceSelectionPolicy.java:
```diff
@@ -0,0 +1,54 @@
+    @NotNull
+    public Iterator<SidecarInstance> iterator()
+    {
+        return instancesProvider.instances().iterator();
```
The order depends in the value returned from `instancesProvider.instances()`, which is implementation dependent. For example, the provider could shuffle the instances and return. In this example, the `OrderedInstanceSelectionPolicy` is no different from `RandomInstanceSelectionPolicy`. 

nit: It might want to sort the instances, so the order is guaranteed.

frankgh replied:
good point. I guess this really means that we won't randomly sort the instances provided

yifan-c replied:
yeah. It feels off semantically. But it is likely not an issue to worry about pragmatically.

frankgh replied:
what I can probably do for now is write a more descriptive javadoc. Let me know if that works out

yifan-c replied:
It is fine to leave it as-is. The original comment is a nit.

yifan-c reviewed (commented):
current progress: ~1/4

yifan-c commented on cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicaProvider.java:
```diff
@@ -0,0 +1,165 @@
+    }
+
+    private List<TokenRangeReplicasResponse.ReplicaInfo>
+    writeReplicasFromPendingRanges(Map<List<String>, List<String>> readReplicaMappings,
```
It is confusing to call it read replicas. We can use the same term, natural, as used in Cassandra code base.
```suggestion
    writeReplicasFromPendingRanges(Map<List<String>, List<String>> naturalReplicaMappings,
```

frankgh replied:
yeah, makes sense, it always felt odd to call it read replicas.

yifan-c commented on cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicaProvider.java:
```diff
@@ -0,0 +1,165 @@
+                                   String keyspace)
+    {
+        LOGGER.debug("Pending token ranges for keyspace={}, pendingRangeMappings={}", keyspace, pendingRangeMappings);
+        // Merge primary and pending replicas to generate candidates for write-replicas
```
```suggestion
        // Merge natural and pending range replicas to generate candidates for write-replicas
```

yifan-c commented on cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicaProvider.java:
```diff
@@ -0,0 +1,165 @@
+                                                               new HashSet<>(entry.getValue())))
+                                                  .collect(Collectors.toList());
+
+        // Candidate write-replica mappings are normalized by removing overlapping ranges
```
It took me some time to finally realize what the `normalize` and `deoverlap` methods do. I think adding the following comments should help future readers. 

```suggestion
        // Candidate write-replica mappings (merged from natural and pending ranges) are normalized by consolidate overlapping ranges
        // For an overlapping range that is included in both natural and pending ranges, say R_natural and R_pending 
        // (where R_natural == R_pending), the replicas of both R_natural and R_pending should receive writes. 
        // Therefore, the write-replicas of such range is the union of both replica sets. 
        // The procedure (esp. normalize method) implements the consolidation process.
```

yifan-c reviewed (commented):
added a few suggestions for `TokenRangeReplicaProvider`

yifan-c commented on client/src/test/java/org/apache/cassandra/sidecar/client/retry/BasicRetryPolicyTest.java:
```diff
@@ -0,0 +1,376 @@
+        assertThat(future.get()).isSameAs(mockResponse);
+    }
+
+    @ParameterizedTest(name = "{index} => canRetryOnADifferentHost={0}")
+    @ValueSource(booleans = { true, false })
+    void testCompletesWithNotFoundStatusCode(boolean canRetryOnADifferentHost)
```
👍

yifan-c commented on client/src/test/java/org/apache/cassandra/sidecar/client/retry/BasicRetryPolicyTest.java:
```diff
@@ -0,0 +1,376 @@
+        assertThat(retryActionCalls.get()).isEqualTo(maxRetries);
+    }
+
+    private static Stream<Arguments> clientStatusCodeArguments()
```
nit: All `4XX` are client error. The stream returned is not the complete set. The original implementation only returns  the established client error codes, which is a subset. That being said, it is rarely a good idea to have custom status code. I only suggest to cover all 4XX code from the perspective of testing, because all of them technically represent client errors, regardless of the predefined or the custom. 

```suggestion
    private static Stream<Arguments> clientStatusCodeArguments()
    {
        return IntStream.range(400, 500).boxed().map(Arguments::of);
    }
```

yifan-c commented on client/src/test/java/org/apache/cassandra/sidecar/client/retry/BasicRetryPolicyTest.java:
```diff
@@ -0,0 +1,376 @@
+        testWithRetries(mockRequest, mockResponse, null, 2, 200, 0, true, false);
+    }
+
+    private static Stream<Arguments> serverStatusCodeArguments()
```
same for serve error, `5xx`

yifan-c commented on client/src/test/java/org/apache/cassandra/sidecar/client/retry/ExponentialBackoffRetryPolicyTest.java:
```diff
@@ -0,0 +1,139 @@
+    void testOverflows()
+    {
+        long delay = new ExponentialBackoffRetryPolicy(1, 1, 0)
+                     .retryDelayMillis(66); // 2 ^ 63 + 1 overflows
```
It should be `64` instead of `66`. `Long.MAX_VALUE == 2 ^ 63`. So to trigger the current overflow check (which is not complete), the input attempts should be 63 + 1 = 64.

The reason why the current check is not complete can be demoed in this test, which fails.

```java
        delay = new ExponentialBackoffRetryPolicy(1, 8, 0)
                .retryDelayMillis(63); // 2 ^ 62 does not overflow, but 2 ^ 62 * 8 overflows. It should cap the value at MAX_VALUE.
        assertThat(delay).isEqualTo(Long.MAX_VALUE);
```

frankgh replied:
updated comment, and added a new test case

yifan-c commented on client/src/test/java/org/apache/cassandra/sidecar/client/retry/RunnableOnStatusCodeRetryPolicyTest.java:
```diff
@@ -0,0 +1,101 @@
+ */
+class RunnableOnStatusCodeRetryPolicyTest
+{
+    private static final Logger LOGGER = LoggerFactory.getLogger(RunnableOnStatusCodeRetryPolicyTest.class);
```
not used.

yifan-c commented on client/src/test/java/org/apache/cassandra/sidecar/client/selection/OrderedInstanceSelectionPolicyTest.java:
```diff
@@ -0,0 +1,79 @@
+        InstanceSelectionPolicy instanceSelectionPolicy = new OrderedInstanceSelectionPolicy(provider);
+        Iterator<SidecarInstance> iterator = instanceSelectionPolicy.iterator();
+
+        assertThat(iterator.hasNext()).isTrue();
+        assertThat(iterator.hasNext()).isTrue().as("Expected to be true");
```
nit: duplication

frankgh replied:
testing idempotency of hasNext. I will add a comment to make it clear

yifan-c commented on client/src/test/java/org/apache/cassandra/sidecar/client/selection/RandomInstanceSelectionPolicyTest.java:
```diff
@@ -0,0 +1,72 @@
+        InstanceSelectionPolicy instanceSelectionPolicy = new RandomInstanceSelectionPolicy(provider);
+        Iterator<SidecarInstance> iterator = instanceSelectionPolicy.iterator();
+
+        assertThat(iterator.hasNext()).isTrue();
+        assertThat(iterator.hasNext()).isTrue().as("Expected to be true");
```
nit: duplication

frankgh replied:
testing idempotency of hasNext. I will add a comment to make it clear

yifan-c commented on client/src/test/java/org/apache/cassandra/sidecar/client/selection/SingleInstanceSelectionPolicyTest.java:
```diff
@@ -0,0 +1,56 @@
+        Iterator<SidecarInstance> iterator = instanceSelectionPolicy.iterator();
+
+        assertThat(iterator.hasNext()).isTrue();
+        assertThat(iterator.hasNext()).isTrue().as("Expected to be true");
```
nit: duplication

frankgh replied:
I'm testing idempotency of hasNext here, so I will add a comment to make it clear

yifan-c commented on client/src/testFixtures/java/org/apache/cassandra/sidecar/client/request/GossipInfoRequestTestParameters.java:
```diff
@@ -0,0 +1,72 @@
+               "\"releaseVersion\":\"4.0.8\",\"hostId\":\"dba02656-ea8c-4a1d-8011-cbc0dab5f411\"," +
+               "\"nativeAddressAndPort\":\"127.0.0.2:9042\",\"load\":\"89084.0\",\"sstableVersions\":\"big-nb\"," +
+               "\"tokens\":\"<hidden>\",\"rpcReady\":\"true\",\"dc\":\"LO\",\"netVersion\":\"12\"," +
+               "\"statusWithPort\":\"NORMAL,-3074457345618258603\"}}[root@b77ac3582e25 /]";
```
`[root@b77ac3582e25 /]`?? 

Btw, the nit on preparing the response body is to build a map and serialize into json. This way, the code reads more structured, instead of a large block of string. Feel free to ignore the nit though.

frankgh replied:
removed the added string, but I agree with you about making the test prep more robust

frankgh replied:
for now, I will keep the clean string.

yifan-c commented on client/src/testFixtures/java/org/apache/cassandra/sidecar/client/request/RingRequestForKeyspaceTestParameters.java:
```diff
@@ -0,0 +1,77 @@
+ */
+public class RingRequestForKeyspaceTestParameters implements RequestTestParameters<RingResponse>
+{
+    public static final String KEYSPACE = "tutorialspoint";
```
hmm... it seems to be the name of a website?

frankgh replied:
it should just be testkeyspace or something like that. I will update it

yifan-c commented on common/src/main/java/org/apache/cassandra/sidecar/common/CassandraAdapterDelegate.java:
```diff
@@ -145,47 +124,73 @@ private void healthCheckInternal()
+                currentVersion = SimpleCassandraVersion.create(releaseVersion);
+                adapter = versionProvider.cassandra(releaseVersion)
+                                         .create(cqlSessionProvider, jmxClient);
+                nodeSettings = newNodeSettings;
```
There is a potential race that set `nodeSettings` to an unexpected value. But since it runs periodically, it should be corrected in the following runs. So.. I think it is acceptable.

frankgh replied:
this shouldn't be an issue because this method is guaranteed to only run by a single thread

yifan-c replied:
yep

yifan-c commented on common/src/main/java/org/apache/cassandra/sidecar/common/data/GossipInfoResponse.java:
```diff
@@ -0,0 +1,292 @@
+
+        // Below are copied from org.apache.cassandra.gms.ApplicationState
+        // Note: that all padding fields are not included.
+        @Deprecated STATUS, //Deprecated and unsued in 4.0, stop publishing in 5.0, reclaim in 6.0
```
```suggestion
        @Deprecated STATUS, //Deprecated and unused in 4.0, stop publishing in 5.0, reclaim in 6.0
```

yifan-c commented on common/src/main/java/org/apache/cassandra/sidecar/common/data/GossipInfoResponse.java:
```diff
@@ -0,0 +1,292 @@
+        {
+            String key = toLowerCamelCase(snakeCaseKey.toUpperCase());
+            super.put(key, value);
+        }
```
How do you make sure the input `String snakeCaseKey` is indeed in snake case. I would suggest to convert the value into enum `GossipField` first by looking up using `valueOf()`, and relocate the `toLowerCamelCase()` to the enum, which simplifies the implementation a bit, since validation for null or empty is no longer required. 

```java
    // in protected enum GossipField
        String toLowerCamelCase()
        {
            String upperUnderscoreCase = name();

            StringBuilder sb = new StringBuilder(upperUnderscoreCase.length());
            sb.append(Character.toLowerCase(upperUnderscoreCase.charAt(0)));

            int length = upperUnderscoreCase.length();
            for (int i = 1; i < length; i++)
            {
                if (upperUnderscoreCase.charAt(i) == '_' && i + 1 < length)
                {
                    i++;
                    sb.append(upperUnderscoreCase.charAt(i));
                }
                else
                {
                    sb.append(Character.toLowerCase(upperUnderscoreCase.charAt(i)));
                }
            }

            return sb.toString();
        }
```

yifan-c reviewed (commented):
Submit the comments so far. Current progress at ~2/5.

yifan-c commented on common/src/test/java/org/apache/cassandra/sidecar/common/SimpleCassandraVersionProviderTest.java:
```diff
@@ -78,7 +78,7 @@ void ensureOutOfOrderInsertionWorks()
                           .add(new V30()).build();
 
-        ICassandraFactory cassandra = provider.getCassandra(SimpleCassandraVersion.create("4.0.0"));
+        ICassandraFactory cassandra = provider.cassandra(SimpleCassandraVersion.create("4.0.0"));
         assertThat(cassandra).hasSameClassAs(new V40());
```
~To test `ensureOutOfOrderInsertionWorks`, should it get V30 from version `SimpleCassandraVersion.create("3.0.0")`?~

~After I made the change, it actually fails, meaning the insertion order has to be in the ascending order of the versions. Either updating the test case or the implementation of builder is needed.~

It is unrelated with this patch, please ignore.

yifan-c commented on src/main/java/org/apache/cassandra/sidecar/cluster/instance/InstanceMetadata.java:
```diff
@@ -50,14 +48,9 @@
-     * @return {@link JmxClient} for connecting with the instance via JMX
-     */
-    JmxClient jmxClient();
+    String uploadsStagingDir();
```
Can you call it `stagingDir`? It is just a directory for the staging data, not necessary to distinguish between scenarios like upload or others.

yifan-c commented on src/main/java/org/apache/cassandra/sidecar/concurrent/ConcurrencyLimiter.java:
```diff
@@ -0,0 +1,112 @@
+public class ConcurrencyLimiter
+{
+    @VisibleForTesting
+    final AtomicInteger permits = new AtomicInteger(0);
```
instead of exposing the mutable `AtomicInteger`, can you add a method to return the current `permits: int` and make the filed private instead? 

```java
    public int acquiredPermits()
    {
        return permits.get();
    }
```

yifan-c commented on src/main/java/org/apache/cassandra/sidecar/data/SchemaRequest.java:
```diff
@@ -15,24 +15,34 @@
  * request parameters
  */
-public class SchemaRequest extends QualifiedTableName
+public class SchemaRequest
```
Should we rename this group of classes to Payload in this patch? For this one, `SchemaRequestPayload`

frankgh replied:
this is not really the payload, these are path params.

yifan-c commented on src/main/java/org/apache/cassandra/sidecar/routes/FileStreamHandler.java:
```diff
@@ -24,54 +24,60 @@
-    private static final Logger LOGGER = LoggerFactory.getLogger(FileStreamHandler.class);
     public static final String FILE_PATH_CONTEXT_KEY = "fileToTransfer";
-
+    private static final Logger LOGGER = LoggerFactory.getLogger(FileStreamHandler.class);
```
logger defined in base class already

yifan-c commented on src/main/java/org/apache/cassandra/sidecar/routes/SnapshotsHandler.java:
```diff
@@ -0,0 +1,364 @@
+    protected SnapshotRequest extractParamsOrThrow(final RoutingContext context)
+    {
+        boolean includeSecondaryIndexFiles =
+        "true".equalsIgnoreCase(context.request().getParam(INCLUDE_SECONDARY_INDEX_FILES, "false"));
```
nit
```suggestion
        boolean includeSecondaryIndexFiles = Boolean.parseBoolean(context.request().getParam(INCLUDE_SECONDARY_INDEX_FILES, "false"));
```

yifan-c commented on src/main/java/org/apache/cassandra/sidecar/routes/sstableuploads/SSTableCleanupHandler.java:
```diff
@@ -0,0 +1,110 @@
+                               SocketAddress remoteAddress,
+                               String uploadId)
+    {
+        if (context.request().method() != HttpMethod.DELETE)
+        {
+            context.fail(wrapHttpException(HttpResponseStatus.BAD_REQUEST, "Unknown API requested"));
+            return;
+        }
```
Why would the request router route such request? It is only reregistered for `delete`

frankgh replied:
oh interesting. that should not be there.

yifan-c commented on src/main/java/org/apache/cassandra/sidecar/utils/BaseFileSystem.java:
```diff
@@ -0,0 +1,125 @@
+import org.apache.cassandra.sidecar.concurrent.ExecutorPools;
+
+/**
+ * Provides functionality for filesytem operations
```
```suggestion
 * Provides functionality for filesystem operations
```

yifan-c commented on src/main/java/org/apache/cassandra/sidecar/utils/CacheFactory.java:
```diff
@@ -0,0 +1,101 @@
+                       .removalListener((RemovalListener<SSTableImporter.ImportOptions, Future<Void>>)
+                                        (options, result, cause) -> {
+                                            LOGGER.debug("Removed entry '{}' with options '{}' from SSTable Import " +
+                                                         "Cache  and cause {}", result, options, cause);
```
```suggestion
                                                         "Cache and cause {}", result, options, cause);
```

yifan-c commented on src/main/java/org/apache/cassandra/sidecar/utils/MD5ChecksumVerifier.java:
```diff
@@ -0,0 +1,101 @@
+
+        Promise<String> result = Promise.promise();
+        file.pause()
+            .setReadBufferSize(64 * 1024)
```
Declare a constant for the buffer size?

yifan-c commented on src/main/java/org/apache/cassandra/sidecar/utils/RequestUtils.java:
```diff
@@ -18,30 +18,29 @@
-            return host.startsWith("[") ? host.substring(1, host.length() - 1) : host;
-        }
-        return address;
+        String value = request.getParam(headerName);
+        if ("true".equalsIgnoreCase(value))
+            return true;
+        if ("false".equalsIgnoreCase(value))
+            return false;
+        return defaultValue;
```
Use `Boolean.parseBoolean()` instead?

frankgh replied:
it's not quite the same. This method takes only true or false, case insensitively. parseBoolean compares to true only.

yifan-c replied:
I thought it is equivalent to 
```java
        return Boolean.parseBoolean(request.getParam(headerName, String.valueOf(defaultValue)));
```

yifan-c commented on src/main/java/org/apache/cassandra/sidecar/utils/SSTableImporter.java:
```diff
@@ -0,0 +1,587 @@
+            isQueueInUse.set(false);
+        }
+
+        public boolean removeIf(Predicate<? super Pair<Promise<Void>, ImportOptions>> filter)
+        {
+            return super.removeIf(filter);
+        }
```
Remove it?

yifan-c commented on src/test/integration/org/apache/cassandra/sidecar/IntegrationTestBase.java:
```diff
@@ -0,0 +1,131 @@
+
+/**
+ * Base class for integration test.
+ * Start a docker container of cassandra at the begining of each test, and
```
```suggestion
 * Start a docker container of cassandra at the beginning of each test, and
```

frankgh committed:
Address PR feedback from Yifan

yifan-c commented on vertx-client-shaded/build.gradle:
```diff
@@ -0,0 +1,123 @@
+ * reference shaded class names.
+ */
+@CacheableTransformer
+class NettyResourceTransformer implements Transformer {
```
MAGIC!

frankgh replied:
this is just renaming netty directories in the fat jar :)

yifan-c reviewed (commented):
Last batch done

frankgh committed:
Address PR feedback from Yifan

frankgh committed:
Address PR feedback from Yifan

frankgh committed:
Address PR feedback from Yifan

frankgh commented:
Thanks for the review @yifan-c . I've pushed a few commits and I hope I've addressed your comments fully. Let me know if there is anything additional to address

yifan-c reviewed (commented):
LGTM. I will take a second pass of the token range replica changes tomorrow.

yifan-c commented on cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicas.java:
```diff
@@ -0,0 +1,515 @@
+    {
+        validateMatchingPartitioner(other);
+
+        BigInteger maxValue = this.partitioner.maxToken;
+        if (this.start.compareTo(other.start) == 0)
+        {
+            if (this.end.equals(maxValue)) return 1;
+            else if (other.end.equals(maxValue)) return -1;
+            else return this.end.compareTo(other.end);
+        }
+        else return this.start.compareTo(other.start);
```
It looks wrong to me..

Say R1 (1, MAX] and R2 (1, MAX], apparently R1 == R2, but according to this implement `R1.compareTo(R2) == 1`

yifan-c commented on cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicas.java:
```diff
@@ -0,0 +1,515 @@
+            return false;
+        }
+        TokenRangeReplicas that = (TokenRangeReplicas) o;
+        return (this.start.equals(that.start) && this.end.equals(that.end) && this.partitioner == that.partitioner);
```
Use `Objects.equals` instead..

yifan-c commented on cassandra40/src/main/java/org/apache/cassandra/sidecar/cassandra40/TokenRangeReplicas.java:
```diff
@@ -0,0 +1,515 @@
+    @Override
+    public int hashCode()
+    {
+        return Objects.hashCode(start, end, partitioner);
```
`Objects.hash`

frankgh committed:
Remove the Token Range Replicas handler implementation

frankgh committed:
Run integration tests in circle

frankgh committed:
Fix MIN_FREE_SPACE_PERCENT_FOR_UPLOAD property (sstable_uploads -> sstable_upload)

dineshjoshi reviewed (approved):
+1

frankgh commented:
Closed via https://github.com/apache/cassandra-sidecar/commit/38cdacb2e7418e2aefbcffb1754dcd324c46028d
